### PR TITLE
Add NamespacedKeyword::is_reverse and NamespacedKeyword::to_reverse f…

### LINKED ISSUE #202: Expose is_reversed, to_reversed on NamespacedKeyword
A keyword naming an attribute — `:person/name`, `:person/friend` — can be prefixed with `_` to denote that it runs backwards. For example, the following two snippets are identical:

```
[?y :person/friend ?x]
[?x :person/hired ?y]
```
```
[?y :person/friend ?x]
[?y :person/_hired ?x]
```

In the query and transaction parsers it'll be useful to define checker and transformation methods on `NamespacedKeyword` to detect and switch between these cases.

### PR DESCRIPTION
**Pull request by jsantell**
…or query and transaction parsers. Fixes #202

Notes: Should we change all of the symbols to be able to handle taking a `String` instead of just a `&str`? Is there a less ugly concat way to handle `to_reverse`?  Should `to_reverse` flip ":foo/_baz" back to ":foo/baz", and if so, maybe a name change to `invert`?

### CODE REVIEW

**Comment by rnewman** [comment]
That is: a symbol's name never starts with '_' for our purposes. I'm not sure if that's true or not in edn as a whole, so pushback is acceptable. Perhaps we want a wrapper struct…

**Comment by jsantell** [comment]
@rnewman makes way more sense! and in this case, to_reverse seems like it should always be able to reverse an already 'reversed' keyword?

**Comment by jsantell** [comment]
Made changes suggested -- should it be `[to|is]_reverse` or `[to|is]_reversed`? Can land this on top in `fmt` in #209

**Comment by jsantell** [comment]
Rebased on #209

File: edn/src/symbols.rs [hunk 1]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
 }
 
 impl NamespacedKeyword {
+    /// Creates a new `NamespacedKeyword`. When `name` starts with a `_`, it is considered
```
Comment (ncalexan) [inline edn/src/symbols.rs]: nit: double newline after the first sentence.

Consider repeating the `is_reversed` notes, with an example, here.

File: edn/src/symbols.rs [hunk 2]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
...
+    /// # use edn::symbols::NamespacedKeyword;
+    /// let nsk = NamespacedKeyword::new("foo", "bar");
+    /// assert!(!nsk.is_reversed());
+    /// assert_eq!(":foo/bar", nsk.to_string());
+    ///
+    /// let reversed = nsk.to_reversed();
+    /// assert!(reversed.is_reversed());
+    /// assert_eq!(":foo/_bar", reversed.to_string());
+    /// ```
+    pub fn to_reversed(&self) -> NamespacedKeyword {
```
Comment (ncalexan) [inline edn/src/symbols.rs]: Maybe `reverse`?

File: edn/src/symbols.rs (hunk 1 above)
Comment (ncalexan) [inline edn/src/symbols.rs]: DataScript and the Clojure implementation refer to "not reversed" as _straight_.  Are we happy using this terminology here?  (I could imagine picking something that doesn't have any connection to sexual orientation.)  If we're happy, include _straight_ as the opposite of reversed in the comment, and add `is_straight` and `straight`/`to_straight`/`straighten`.

The fact that _straight_ and _reverse_ conjugate so differently argues for changing _straight_ but I don't have a second suggestion to hand.

File: edn/src/symbols.rs [hunk 3]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
...
+        NamespacedKeyword {
```
Comment (ncalexan) [inline edn/src/symbols.rs]: We are really making things hard to pattern match against, aren't we?  Oh well, don't have a better suggestion right now.

**Comment by rnewman** [comment]
> I think this interface will prove awkward in practice, and doesn't help folks who need to consider reversed attributes (the transactor) actually ensure they handled the reversed cases, but I don't have a better approach to hand.

I was expecting something like:

```
let attribute = …;    // Parser output
let reversed = attribute.is_reversed();
return Pattern {
  entity: if reversed { value } else { entity },
  attribute: if reversed { attribute.reverse() } else { attribute },
  value: if reversed { entity } else { value },
  tx: tx,
};
```

which is code I'm pretty much writing now.

File: edn/src/symbols.rs [hunk 4]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
...
         assert!(!namespace.is_empty(), "Keywords cannot have an empty non-null namespace.");
+        let reversed = match name.chars().nth(0) {
+            Some(v) => v.to_string() == "_",
+            None => false,
+        };
 
         // TODO: debug asserts to ensure that neither field matches [ :/].
-        NamespacedKeyword { name: name.to_string(), namespace: namespace.to_string() }
+        NamespacedKeyword {
+            name: if reversed { name[1..].to_string() } else { name.to_string() },
```
Comment (rnewman) [inline edn/src/symbols.rs]: This needs to be rebased on top of #213.

File: edn/src/symbols.rs [hunk 5]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
...
+    ///
+    /// ```rust
+    /// # use edn::symbols::NamespacedKeyword;
+    /// let keyword = NamespacedKeyword::new("foo", "bar");
+    /// assert_eq!(keyword.to_string(), ":foo/bar");
+    /// ```
     pub fn new(namespace: &str, name: &str) -> Self {
         assert!(!name.is_empty(), "Keywords cannot be unnamed.");
         assert!(!namespace.is_empty(), "Keywords cannot have an empty non-null namespace.");
+        let reversed = match name.chars().nth(0) {
```
Comment (rnewman) [inline edn/src/symbols.rs]: Use `name.starts_with('_')`.

File: edn/src/symbols.rs [hunk 6]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
...
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use edn::symbols::NamespacedKeyword;
+    /// assert!(!NamespacedKeyword::new("foo", "bar").is_reversed());
+    /// assert!(NamespacedKeyword::new("foo", "_bar").is_reversed());
+    /// assert_eq!(NamespacedKeyword::new("foo", "_bar").name, "bar");
+    /// ```
+    pub fn is_reversed(&self) -> bool {
```
Comment (rnewman) [inline edn/src/symbols.rs]: `#[inline]`

File: edn/src/symbols.rs (hunk 2 above)
Comment (rnewman) [inline edn/src/symbols.rs]: I think `reverse` would imply mutability. And while I'd normally be happy with mutability in this case, the keywords we're working with will probably be borrowed references inside values.

File: edn/src/symbols.rs [hunk 7]
```diff
@@ -98,12 +99,73 @@ impl Keyword {
...
+    /// assert_eq!(":foo/bar", nsk.to_string());
+    ///
+    /// let reversed = nsk.to_reversed();
+    /// assert!(reversed.is_reversed());
+    /// assert_eq!(":foo/_bar", reversed.to_string());
+    /// ```
+    pub fn to_reversed(&self) -> NamespacedKeyword {
+        NamespacedKeyword {
+            name: self.name.clone(),
+            namespace: self.namespace.clone(),
```
Comment (rnewman) [inline edn/src/symbols.rs]: There's 'spread' syntax for this:

```
NamespacedKeyword { reversed: !self.reversed, .. self }
```

File: edn/src/symbols.rs [hunk 8]
```diff
@@ -161,8 +223,9 @@ impl Display for NamespacedKeyword {
     /// ```rust
     /// # use edn::symbols::NamespacedKeyword;
     /// assert_eq!(":bar/baz", NamespacedKeyword::new("bar", "baz").to_string());
+    /// assert_eq!(":bar/_baz", NamespacedKeyword::new("bar", "_baz").to_string());
```
Comment (rnewman) [inline edn/src/symbols.rs]: Also add a `.to_reversed()` test here. You could make this new assertion pass without changing any of the other code in the file!

File: edn/src/symbols.rs (hunk 1 above)
Comment (rnewman) [inline edn/src/symbols.rs]: I'm not super happy with 'straight'. Quite apart from the exclusionary language, this isn't a thing that bends; it's a directional flag where the two directions are opposite. So `reversed`, or `direction: forwards/backwards`, perhaps even `inverse`, but not `straight`.

File: edn/src/symbols.rs (hunk 5 above)
Comment (jsantell) [inline edn/src/symbols.rs]: Lovely.

File: edn/src/symbols.rs [hunk 9]
```diff
@@ -101,14 +101,89 @@ impl Keyword {
...
+    /// assert_eq!(":foo/_bar", reversed.to_string());
+    /// ```
+    pub fn to_reverse(&self) -> NamespacedKeyword {
+        let name = if self.is_backward() {
+            self.name[1..].to_string()
+        } else {
+            format!("{}{}", "_", self.name)
+        };
+
+        NamespacedKeyword {
```
Comment (jsantell) [inline edn/src/symbols.rs]: Unfortunately, spread properties don't seem to work due to needing to clone the props

File: edn/src/symbols.rs [hunk 10]
```diff
@@ -101,14 +101,89 @@ impl Keyword {
...
+    ///
+    /// let reversed = nsk.to_reverse();
+    /// assert!(reversed.is_backward());
+    /// assert_eq!(":foo/_bar", reversed.to_string());
+    /// ```
+    pub fn to_reverse(&self) -> NamespacedKeyword {
+        let name = if self.is_backward() {
+            self.name[1..].to_string()
+        } else {
+            format!("{}{}", "_", self.name)
```
Comment (jsantell) [inline edn/src/symbols.rs]: Better way?

Comment (ncalexan) [inline edn/src/symbols.rs]: `concat`, `+`?  I don't much care, this is clear.

File: edn/src/symbols.rs [hunk 11]
```diff
@@ -166,6 +241,7 @@ impl Display for NamespacedKeyword {
     /// ```rust
     /// # use edn::symbols::NamespacedKeyword;
     /// assert_eq!(":bar/baz", NamespacedKeyword::new("bar", "baz").to_string());
+    /// assert_eq!(":bar/_baz", NamespacedKeyword::new("bar", "baz").to_reverse().to_string());
```
Comment (ncalexan) [inline edn/src/symbols.rs]: Can you verify that `to_reverse().to_reverse()` is the identity?

Comment (jsantell) [inline edn/src/symbols.rs]: :+1:

**Pull request merged**